File: 数学/leetcode29.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        """思路：官方跑题有问题"""
        if(not dividend): return 0
        minus = (dividend>0)^(divisor>0)
        dividend = abs(dividend)
        divisor = abs(divisor)
        ans = 0
        now_divisor = divisor
        k_lists = []
        divisor_lists = []
        k = 1
        while(dividend>=now_divisor):
            if(now_divisor>dividend): break
            k_lists.append(k)
            k += k
            divisor_lists.append(now_divisor)
            now_divisor += now_divisor
        while(dividend>=divisor):
            k = k_lists.pop()
            now_divisor = divisor_lists.pop()
            if(dividend>now_divisor):
                ans += k
                dividend -= now_divisor
            elif(dividend==now_divisor):
                ans += k
                break
        return -ans if minus else ans

File: math/leetcode29.cpp
#include <iostream>
using namespace std;
class Solution {
public:
    int divide(int dividend, int divisor) {
        if(dividend==0) return 0;
        int flag = (dividend>0)^(divisor>0);
        dividend = dividend>0?dividend:-dividend;
        divisor = divisor>0?divisor:-divisor;
        int ans = minus(dividend,divisor);
        return flag?-ans:ans;
    }
    int minus(int dividend, int divisor){
        int k = 0;
        int now_divisor = divisor;
        while(dividend>now_divisor){
            k++;
            now_divisor <<= 1;
        }
        int ret = 0;
        while(dividend>=divisor){
            if(dividend>now_divisor){
                ret += 2<<k;
                dividend -= now_divisor;
            }else if(dividend==now_divisor){
                ret += 2<<k;
                return ret;
            }
            k--;
            now_divisor >>= 1;
        }
        return ret;
    }
};
int main()
{
    Solution s;
    int ans = s.divide(10,3);
    cout << ans;
}
